### PR TITLE
feat(typescript): Inherit types from `defaults.Octokit` set via `App.defaults({ Octokit })`

### LINKED ISSUE #226: Inherit types from `defaults.Octokit` set via `App.defaults({ Octokit })`
See https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/issues/32

### PR DESCRIPTION
**Pull request by gr2m**
fixes #226

This is work in progress, I didn't get it to work yet. I'm not sure why

### CODE REVIEW

**Comment by gr2m** [comment]
The problem is not the `Option` type, but `defaults: Partial<TDefaults>`.

When passing valid options (with all required keys), it works as expected:

```ts
const MyApp = App.defaults({ Octokit: MyOctokit, appId: "", privateKey: "" });
const app = new MyApp({
  appId: APP_ID,
  privateKey: PRIVATE_KEY,
});

app.octokit.foo;
// typed as string
```

But it doesn't if either `appId` or `privateKey` is missing

```ts
const MyApp = App.defaults({ Octokit: MyOctokit });
const app = new MyApp({
  appId: APP_ID,
  privateKey: PRIVATE_KEY,
});

app.octokit.foo;
// Error: Property 'foo' does not exist on type 'Octokit
```

**Comment by wolfy1339** [comment]
Are `appID` and `privateKey` always required?

You can do this if they are:
```ts
default: Partial<TDefaults> & { appID: number; privateKey: string }
```

**Comment by gr2m** [comment]
> Are `appID` and `privateKey` always required?
> 
> You can do this if they are:
> 
> ```ts
> default: Partial<TDefaults> & { appID: number; privateKey: string }
> ```

that's exactly what I just did, and it works great! Thanks a lot @wolfy1339, next time I'll ask you for help before I break my brain

File: src/types.ts [hunk 1]
```diff
@@ -21,6 +21,14 @@ export type Options = {
   };
 };
 
+// workaround for https://github.com/octokit/app.js/pull/227
+// we cannot make appId & privateKey required on Options because
+// it would break inheritance of the Octokit option set via App.defaults({ Octokit })
+export type ConstructorOptions<TOptions extends Options> = TOptions & {
+  appId: number | string;
+  privateKey: string;
+};
```
Comment (gr2m) [inline src/types.ts:30]: @wolfy1339 this was the significant change that made it all work 🎉

**Pull request merged**